astropy/cosmology/core.py

```diff
@@ -96,15 +96,15 @@ class Cosmology(metaclass=abc.ABCMeta):
     write = UnifiedReadWriteMethod(CosmologyWrite)
 
     # Parameters
-    parameters = ParametersAttribute()
+    parameters = ParametersAttribute(attr_name="_parameters")
```

Comment:
Don't really have time for a full review, but why the new `attr_name`? It seems unrelated to the PR and not really used.

Reply:
It's because the current form assumes the ``ParametersAttribute`` is connected to an underscore-prefixed class attribute that's created in `Cosmology.__init_subclass__`. For a private attribute that assumption means the connected attribute is double-underscore-prefixed! This is of course a problem because attribute access is different for double-underscore-prefixed names and ``getattr(instance, "__derived_paramters")`` doesn't work. 

This PR changes ``__set_name__` from building ``attr_name`` based upon this problematic assumption to having ``attr_name`` be set at ``ParametersAttribute`` construction so as to avoid double-underscore-prefixed attributes. In the next line (100) you'll see that `_derived_parameters` is now connected to `_parameters_derived` (not `__derived_parameters`), which is the dict of derived `Parameter` objects and the counterpart to `_parameters` in this line.

Reply:
Ah, fair enough! Though why not use the same trick as `lazyproperty` and just store the actual data under the same name in `parent.__dict__`?

Reply:
Does that work for class attributes?

Reply:
Unfortunately this doesn't appear to be viable

![CleanShot 2023-10-29 at 17 52 57@2x](https://github.com/astropy/astropy/assets/8949649/870c347f-dcf2-4b5c-aeee-c9c889861bb5)

Reply:
The only relevant part of the above image is that `__init_subclass__` can add attributes, but not modify the ``__dict__`` since this is a class-level attribute.

Reply:
Unfortunately this doesn't appear to be viable because this is a class variable not an instance variable, like `lazyproperty` works with.

![CleanShot 2023-10-29 at 18 00 59@2x](https://github.com/astropy/astropy/assets/8949649/b4054719-28ee-438d-bdb5-054c38b90a48)

Reply:
No, indeed, that would work only on instances... Sorry for the noise!

Reply:
I think it would work with a metaclass, but let's avoid that  😆

Reply:
I don't love the parameter name `attr_name`. Any suggestions? Maybe `get_attr`...